lib/ipip.c

```diff
@@ -25,34 +25,27 @@
 #include "gatekeeper_net.h"
 
 int
-encapsulate(struct rte_mbuf *pkt, uint8_t priority,
-	struct ipip_tunnel_info *info)
+encapsulate(struct rte_mbuf *pkt, uint8_t priority, struct ip_flow *flow)
 {
-	struct ether_hdr *new_eth;
 	struct ipv4_hdr *outer_ip4hdr;
 	struct ipv6_hdr *outer_ip6hdr;
 
-	if (info->flow.proto == ETHER_TYPE_IPv4) {
+	/*
+	 * The original packet should contain the Ethernet header,
+	 * while the function shouldn't add an Ethernet header.
+	 * When allocating space for the outer IP header,
+	 * it only needs to allocate the extra needed space.
+	 */
```

Comment:
This comment is really good, but it shouldn't be here. It should be in the header file that has the prototype of `encapsulate()` because everyone calling `encapsulate()` should know that.